dashboard/src/api/auth-manager.ts

```diff
@@ -0,0 +1,138 @@
+    public maxTokenAge: number = 5 * 60 * 1000;  // 5 minutes
+
+    private _api: Api;
+    private _axios: axios.AxiosInstance;
+    private _authInterceptorId: number | null = null;
+    private _tokenRefreshPromise: Promise<AuthToken> | null = null;
+
+    constructor(api: Api, axiosInstance: axios.AxiosInstance) {
+        this._api = api;
+        this._axios = axiosInstance;
+    }
+
+    checkExistingLogin(): Promise<AuthToken|null> {
+        return this.refreshLogin().then((authToken) => {
+            if (authToken) {
+                this._mountAxiosInterceptor();
+            }
+            return authToken;
+        });
+    }
+
+    initiateLogin(username: string, password: string): Promise<CodeToken> {
+        return this._axios.post<CodeToken>(
+            this._api.endpoints.authCodeToken, {username, password}).then(
+                (response) => response.data);
+    }
+
+    continueLogin(
+        codeToken: string, verificationCode: string
+    ): Promise<AuthToken> {
+        const authTokenUrl = this._api.endpoints.authAuthToken;
+        const params = {code_token: codeToken, code: verificationCode};
+        return this._axios.post<AuthToken>(authTokenUrl, params).then(
+            (response) => {
+                tokenStorage.storeToken(response.data.token);
+                this._ejectAxiosInterceptor();
+                this._mountAxiosInterceptor();
+                return response.data;
+            },
+            (error) => {
+                tokenStorage.clearToken();
+                throw error;
+            });
+    }
+
+    logout() {
+        tokenStorage.clearToken();
+        this._ejectAxiosInterceptor();
+    }
+
+    refreshLogin(): Promise<AuthToken|null> {
+        if (tokenStorage.getToken()) {
+            return this._refreshToken();
+        }
+        return Promise.resolve(null);
+    }
+
+    private _mountAxiosInterceptor = () => {
+        this._authInterceptorId = (
+            this._axios.interceptors.request.use(this._authRequestInterceptor));
+    }
+
+    private _ejectAxiosInterceptor = () => {
+        if (this._authInterceptorId !== null) {
+            this._axios.interceptors.request.eject(this._authInterceptorId);
+            this._authInterceptorId = null;
+        }
+    }
+
+    private _authRequestInterceptor = (request: axios.AxiosRequestConfig) => {
+        if ((tokenStorage.getTokenAge() || 0) > this.maxTokenAge) {
+            this._refreshToken();
+        }
+        return {
+            ...request,
+            headers: {
+                ...request.headers,
+                'Authorization': `JWT ${tokenStorage.getToken()}`,
+            }
+        };
+    }
+
+    private _refreshToken: (() => Promise<AuthToken>) = () => {
+        if (this._tokenRefreshPromise) {
+            return this._tokenRefreshPromise;
+        }
+        const refreshUrl = this._api.endpoints.authRefresh;
+        const promise = this._tokenRefreshPromise = this._axios.post(
```

Comment:
Reassignment to unneeded variable, a `return this._tokenRef...` would be enough here.

Reply:
This is because I thought that `this._tokenRef..` could be changed to null while prosessing the promise and I wasn't sure if it's possible that the Promise is already resolved (which would set it to null) at line 136 which returns the promise.

Reply:
Alright... I'll allow, but this is a bit hard to understand and the reassignment does not help here.